haskell-best-practices.md

```diff
@@ -184,6 +184,165 @@ instance Monoid Mult where
   mappend = (*)
 ```
 
+## Phantom types
+
+The `Tagged` example above makes use of a phantom type variable. The definition of `Tagged` is:
+```haskell
+newtype Tagged t a = Tagged { untag :: a }
+```
+
+Note how the type variable `t` does not appear on the right-hand-side of the equals-sign. It's only used to add type-level information to a type - it's never used at the value level. We call this kind of type variable a phantom type variable, or occasionally, just a phantom type.
+
+A more useful example is hashing passwords. We might want to represent passwords as either raw text or hashed text and have operations that only work on one kind:
+
+```haskell
+module Password (Password, PasswordState, rawPassword, hashPassword, comparePassword) where
+
+data PasswordState = Raw | Hashed
+
+newtype Password (s :: PasswordState) = Password Text
+
+rawPassword :: Text -> Password 'Raw
+rawPassword = Password
+
+hashPassword :: Password 'Raw -> Password 'Hashed
+hashPassword (Password r) = Password (hash r)
```

Comment:
might want to clarify somewhere where `hash` comes from so people don't go around looking for it?